File: apps/cli/src/config/defaultConfig.ts
/** TypeScript view of `.brain/config.yml`. Mirrors the Python BrainConfig. */

export interface BrainConfigShape {
  brain: { repo_root: string };
  daemon: {
    host: string;
    port: number;
    auto_start: boolean;
    python_command: string;
  };
  indexing: {
    debounce_ms: number;
    max_cpu_cores: number;
    initial_index_parallelism: number;
    incremental_index_parallelism: number;
    batch_size_files: number;
    batch_sleep_ms: number;
    pause_when_system_cpu_above: number;
    pause_when_memory_above_mb: number;
    pause_during_build: boolean;
    pause_when_on_battery: boolean;
  };
  hash_scan: {
    enabled: boolean;
    interval_seconds: number;
    max_files_per_batch: number;
    batch_sleep_ms: number;
  };
  embedding: {
    provider: string;
    model: string;
    batch_size_chunks: number;
    max_chunk_tokens: number;
    only_embed_changed_chunks: boolean;
    lazy_load_model: boolean;
    unload_after_idle_minutes: number;
    base_url: string;
    api_key_env: string;
  };
  retrieval: {
    max_context_tokens: number;
    lexical_top_k: number;
    vector_top_k: number;
    symbol_top_k: number;
    rerank_top_k: number;
    include_tests: boolean;
    include_callers: boolean;
    include_callees: boolean;
    include_config: boolean;
    include_git_diff: boolean;
  };
  ai: {
    provider: string;
    base_url: string;
    api_key_env: string;
    model: string;
    temperature: number;
  };
  apply: {
    require_patch_approval: boolean;
    run_tests_after_apply: boolean;
    test_command: string;
  };
}

export function defaultConfig(): BrainConfigShape {
  return {
    brain: { repo_root: "." },
    daemon: {
      host: "127.0.0.1",
      port: 8765,
      auto_start: true,
      python_command: "python",
    },
    indexing: {
      debounce_ms: 2500,
      max_cpu_cores: 1,
      initial_index_parallelism: 2,
      incremental_index_parallelism: 1,
      batch_size_files: 10,
      batch_sleep_ms: 250,
      pause_when_system_cpu_above: 75,
      pause_when_memory_above_mb: 2048,
      pause_during_build: true,
      pause_when_on_battery: true,
    },
    hash_scan: {
      enabled: true,
      interval_seconds: 600,
      max_files_per_batch: 200,
      batch_sleep_ms: 500,
    },
    embedding: {
      provider: "mock",
      model: "local-code-embedding-small",
      batch_size_chunks: 16,
      max_chunk_tokens: 800,
      only_embed_changed_chunks: true,
      lazy_load_model: true,
      unload_after_idle_minutes: 15,
      base_url: "",
      api_key_env: "OPENAI_API_KEY",
    },
    retrieval: {
      max_context_tokens: 25000,
      lexical_top_k: 20,
      vector_top_k: 30,
      symbol_top_k: 20,
      rerank_top_k: 12,
      include_tests: true,
      include_callers: true,
      include_callees: true,
      include_config: true,
      include_git_diff: true,
    },
    ai: {
      provider: "openai-compatible",
      base_url: "",
      api_key_env: "OPENAI_API_KEY",
      model: "gpt-4.1",
      temperature: 0.1,
    },
    apply: {
      require_patch_approval: true,
      run_tests_after_apply: true,
      test_command: "",
    },
  };
}

File: brain/config/default_config.py
"""Default configuration model for Local Code Brain.

Mirrors the ``.brain/config.yml`` structure described in the product spec.
Pydantic provides validation and defaults so a missing/partial config still
produces a fully-populated, valid configuration.
"""

from __future__ import annotations

from pydantic import BaseModel, ConfigDict, Field
from pydantic.alias_generators import to_camel


class BrainSection(BaseModel):
    model_config = ConfigDict(alias_generator=to_camel, populate_by_name=True)

    repo_root: str = "."


class DaemonSection(BaseModel):
    model_config = ConfigDict(alias_generator=to_camel, populate_by_name=True)

    host: str = "127.0.0.1"
    port: int = 8765
    auto_start: bool = True
    python_command: str = "python"


class IndexingSection(BaseModel):
    model_config = ConfigDict(alias_generator=to_camel, populate_by_name=True)

    debounce_ms: int = 2500
    max_cpu_cores: int = 1
    initial_index_parallelism: int = 2
    incremental_index_parallelism: int = 1
    batch_size_files: int = 10
    batch_sleep_ms: int = 250
    pause_when_system_cpu_above: int = 75
    pause_when_memory_above_mb: int = 2048
    pause_during_build: bool = True
    pause_when_on_battery: bool = True


class HashScanSection(BaseModel):
    model_config = ConfigDict(alias_generator=to_camel, populate_by_name=True)

    enabled: bool = True
    interval_seconds: int = 600
    max_files_per_batch: int = 200
    batch_sleep_ms: int = 500


class EmbeddingSection(BaseModel):
    model_config = ConfigDict(alias_generator=to_camel, populate_by_name=True)

    provider: str = "mock"
    model: str = "local-code-embedding-small"
    batch_size_chunks: int = 16
    max_chunk_tokens: int = 800
    only_embed_changed_chunks: bool = True
    lazy_load_model: bool = True
    unload_after_idle_minutes: int = 15
    # Optional OpenAI-compatible embedding endpoint
    base_url: str = ""
    api_key_env: str = "OPENAI_API_KEY"


class RetrievalSection(BaseModel):
    model_config = ConfigDict(alias_generator=to_camel, populate_by_name=True)

    max_context_tokens: int = 25000
    lexical_top_k: int = 20
    vector_top_k: int = 30
    symbol_top_k: int = 20
    rerank_top_k: int = 12
    include_tests: bool = True
    include_callers: bool = True
    include_callees: bool = True
    include_config: bool = True
    include_git_diff: bool = True


class AiSection(BaseModel):
    model_config = ConfigDict(alias_generator=to_camel, populate_by_name=True)

    provider: str = "none"
    base_url: str = ""
    api_key_env: str = ""
    model: str = ""
    temperature: float = 0.1


class ContextVerifierSection(BaseModel):
    model_config = ConfigDict(alias_generator=to_camel, populate_by_name=True)

    enabled: bool = False
    provider: str = "openrouter"
    base_url: str = "https://openrouter.ai/api/v1"
    model: str = "qwen/qwen3-coder:free"
    api_key_env: str = "OPENROUTER_API_KEY"
    max_attempts: int = 3
    min_confidence: float = 0.75
    explain_with_verifier: bool = True


class ApplySection(BaseModel):
    model_config = ConfigDict(alias_generator=to_camel, populate_by_name=True)

    require_patch_approval: bool = True
    run_tests_after_apply: bool = True
    test_command: str = ""


class BrainConfig(BaseModel):
    model_config = ConfigDict(alias_generator=to_camel, populate_by_name=True)

    brain: BrainSection = Field(default_factory=BrainSection)
    daemon: DaemonSection = Field(default_factory=DaemonSection)
    indexing: IndexingSection = Field(default_factory=IndexingSection)
    hash_scan: HashScanSection = Field(default_factory=HashScanSection)
    embedding: EmbeddingSection = Field(default_factory=EmbeddingSection)
    retrieval: RetrievalSection = Field(default_factory=RetrievalSection)
    ai: AiSection = Field(default_factory=AiSection)
    context_verifier: ContextVerifierSection = Field(default_factory=ContextVerifierSection)
    apply: ApplySection = Field(default_factory=ApplySection)


def default_config() -> BrainConfig:
    """Return a fully-populated default configuration."""

    return BrainConfig()


def default_config_dict() -> dict:
    """Return defaults as a plain dict suitable for YAML serialization."""

    return default_config().model_dump()
